### ISSUE TITLE
TypeScript 5.0 Support

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#6380]
https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-rc/

This issue is just to track all of the new features and their implementation state in this project.
As with all releases, we will not necessarily to support all features until closer to the full release when everything the features are stabilised.

**Please be patient.**

## ✅ Decorators

We'll need to review the new decorators version to double check if there are any syntax differences.

From the looks of it we're missing support for at least this new valid syntax:
```ts
export
@decorator
class Bar {}

export default
@decorator
class Baz {}
```

## ✅ `const` Type Parameters (https://github.com/typescript-eslint/typescript-eslint/pull/6600)

```ts
type HasNames = { names: readonly string[] };
function getNamesExactly<const T extends HasNames>(arg: T): T["names"] {
//                       ^^^^^
    return arg.names;
}
```

This will require AST changes

## ✅ `--verbatimModuleSyntax`

This new flag doesn't actually impact us, but we will want to create a new lint rule to help people write "good" code with this flag on.

With the flag, this is the transpilation behaviour:
```ts
import {type A, type B} from 'mod';

// transpiled to

import {} from 'mod';
// which is the same as
import 'mod';
```
Based on discussions - it's clear that most people won't actually want this behaviour, so we'll want a lint rule that warns against import declarations that only contain inline type qualified specifiers.

#6382 has been raised separately to track this

## ✅ API Breaking Changes

We will need to audit our codebase for breakages and fix them as necessary.

## ✅ `lib.d.ts` Updates

We will need to regenerate our types within `scope-manager`.

-----

## Other changes with no impact to us

- Supporting Multiple Configuration Files in `extends`
- All `enum`s Are Union `enum`s
- `--moduleResolution bundler`
- Resolution Customization Flags
	- `allowImportingTsExtensions`
	- `resolvePackageJsonExports`
	- `resolvePackageJsonImports`
	- `allowArbitraryExtensions`
	- `customConditions`
	- `verbatimModuleSyntax`
- Support for `export type *`
	- This was already syntactically valid and thus our AST already supported this, marking this case as `exportKind: 'type'`. Previously it was a TS semantic error, and it is no more.
- Support in JSDoc
	- `@satisfies`
	- `@overload`
- Passing Emit-Specific Flags Under `--build`
- Exhaustive `switch`/`case` Completions
- Speed, Memory, and Package Size Optimizations
    - This might have implicit downstream wins for lint times.
- Breaking Changes and Deprecations
    - Runtime Requirements
        - We already have minimum node v12, so we're good here.
    - API Breaking Changes
        - We will need to audit our codebase for breakages and fix them as necessary.
    - Forbidden Implicit Coercions in Relational Operators
    - Enum Overhaul
    - More Accurate Type-Checking for Parameter Decorators in Constructors Under `--experimentalDecorators`
    - Deprecations and Default Changes

**Comment by Jack-Works** [comment]
After upgrading to ts 5.0-beta, I found this error:

> This rule requires the `strictNullChecks` compiler option to be turned on to function correctly

Is there some breaks in the config parsing? We use "extends" to inherit strict flag from shared config.

**Comment by bradzacher** [comment]
@Jack-Works https://typescript-eslint.io/maintenance/versioning#typescript
We do not officially support the beta versions, so we have not yet tested our code against the beta.

If something has broken, then that's probably due to an undocumented change in their release, which we'll address before the full 5.0 release.

**Comment by bcowgill** [comment]
I'm not sure about "no impact on us" for "bundler" --moduleResolution setting.

I got this error
swcpack.config.js
  0:0  error  Parsing error: Argument for '--moduleResolution' option must be: 'node', 'classic', 'node16', 'nodenext'

After adding the suggested configuration setting for typescript:
  parserOptions: {
    project: true,
  },
from https://typescript-eslint.io/linting/typed-linting
I set project: false and omitted 
    // 'plugin:@typescript-eslint/recommended-requiring-type-checking'

**Comment by bradzacher** [comment]
@bcowgill please read my previous comment.
https://github.com/typescript-eslint/typescript-eslint/issues/6380#issuecomment-1408133403

**Comment by bradzacher** [comment]
The RC is live https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-rc/

**Comment by sosukesuzuki** [comment]
@bradzacher @JoshuaKGoldberg Are you planning to work on this soon? If you are busy, can I work on this?

**Comment by bradzacher** [comment]
@sosukesuzuki feel free to raise PRs!
